### DISCUSSION TITLE
addition of third Fortran keyword

### DISCUSSION THREAD

**Discussion by nbehrnd** [jgm/pandoc#10099] [Q&A] [answered]
Prior to perhaps prepare a PR, and because I'm not familiar with the Haskel language, I would like to rise the question about a third keyword Pandoc `f90` to recognize snippets of Fortran here.

## observation:

GitHub flavored markdown files allow to fence snippets of code blocks with a tag about the language they represent, for instance

```python
print("hello")
```

or for Fortran

```fortran
program test
   print *, "hello"
end program test
```
 
introduced by a line of with ` ```fortran`

It equally is possible to use a tag ` ```f90`

```f90
program test
   print *, "hello"
end program test
```

on GitHub in the markdown files, issue reports; or e.g. on the discussion board of [fortran-lang.org](https://fortran-lang.discourse.group/). Beside brevity as tag, `f90` as file extension is a typical indicator for the free form syntax since Fortran90 up and including Fortran2023 contrasting to FORTRAN77 / fixed form syntax, too.

Second, I see i) `src/Text/Pandoc/Highlighting.hs` has a `langsList` with the entry `("fortran","Fortran"),` and ii) `src/Text/Pandoc/Writers/MediaWiki.hs` in the block starting by `highlightingLangs = Set.fromList [` the consecutive lines `  "fortran",` and `  "fortranfixed",`.

## question

To provide the third Fortran indicator to pandoc, would it suffice to add `  "f90",` to `src/Text/Pandoc/Writers/MediaWiki.hs` alone?

**Comment by jgm** [comment]
`langsList` there is only for conversion between LaTeX `listings` language names and ours.

What pandoc version, and what format are you converting to?  `f90` should already work.
````
% pandoc  
``` f90
x = 9
^D
```
<div class="sourceCode" id="cb1"><pre
class="sourceCode f90"><code class="sourceCode fortranfree"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>x <span class="kw">=</span> <span class="dv">9</span></span></code></pre></div>
````

**Comment by nbehrnd** [comment]
You are correct; both keywords to fence snippets of Fortran are in place, and both are functional.

However, while editing a draft about object oriented programming in Fortran ([link](https://github.com/nbehrnd/bader_article)), I observed their working differs a little bit.  Eventually, I had to stick to `f90`  throughout the work (their: file [`pandoc_md.md`](https://github.com/nbehrnd/bader_article/blob/main/pandoc_md.md)) for a good representation of the content.

If one copies the following as `snippet.md`

````markdown
Fenced with `f90` -- correct rendering

```f90
TYPE :: sorted_list
   PRIVATE
   TYPE(sortable) :: data
   TYPE(sorted_list), POINTER :: next => null()
CONTAINS
   FINAL :: delete_sorted_list
END TYPE
```

Fenced with `fortran` -- unresolved representation

```fortran
TYPE :: sorted_list
   PRIVATE
   TYPE(sortable) :: data
   TYPE(sorted_list), POINTER :: next => null()
CONTAINS
   FINAL :: delete_sorted_list
END TYPE
```
````

the visual result of the conversion with Pandoc 3.3 differs for either

```bash
pandoc snippet.md -s -o snippet.html
pandoc snippet.md -s -o snippet.pdf -t ms
```

A closer inspection of an intermediate file, for instance the .ms file, reveals 

```bash
$ pandoc snippet.md -s -o snippet.ms
$ grep CONTAINS snippet.ms -n
125:\*[KeywordTok "CONTAINS"]
139:\*[CommentTok "CONTAINS"]
```

once the keyword is resolved _as such_, the other time (erroneously) processed as a comment.  The dictionary accessed by the keyword `f90` thus appears as more abreast to the current set of Fortran keywords, than the one of the lexer by a fence label `fortran`.  Since this actually reflects correctly the history -- because `contains` was introduced as keyword with standard Fortran90 ([reference](https://fortranwiki.org/fortran/show/Keywords)) -- was the lexer launched by `fortran` intentionally frozen to standards up and including FORTRAN77?

---

Addition: I just notice a different visual representation on GitHub, too.   Fenced with `f90`:

```f90
TYPE :: sorted_list
   PRIVATE
   TYPE(sortable) :: data
   TYPE(sorted_list), POINTER :: next => null()
CONTAINS
   FINAL :: delete_sorted_list
END TYPE
```

vs fenced with `fortran`:

```fortran
TYPE :: sorted_list
   PRIVATE
   TYPE(sortable) :: data
   TYPE(sorted_list), POINTER :: next => null()
CONTAINS
   FINAL :: delete_sorted_list
END TYPE
```



[2024-09-06_pandoc_Fortran_fences.zip](https://github.com/user-attachments/files/16909272/2024-09-06_pandoc_Fortran_fences.zip)

**Reply by jgm** [reply, marked as answer]
In skylighting, there are two distinct lexers (derived from KDE's syntax highlighters): fortran-fixed and fortran-free. The latter is used with `f90`, the former with `fortran`.

**Comment by nbehrnd** [comment]
In a thread about syntax lexers on fortran-lang.org, I amended my answer with your information ([link](https://fortran-lang.discourse.group/t/fortran-keyword-categories/1816/3)).  Thank you.
